dali/operators/reader/fits_reader_gpu_op.cu

```diff
@@ -0,0 +1,384 @@
+            }
+            nzero = nbits - nonzero_count[b];
+            nbits -= nzero + 1;
+            b ^= 1 << nbits;
+            nbits -= fs;
+            while (nbits < 0) {
+              b = (b << 8) | (*compressed_tile++);
+              nbits += 8;
+            }
+            diff = (nzero << fs) | (b >> nbits);
+            b &= (1 << nbits) - 1;
+
+            if ((diff & 1) == 0) {
+              diff = diff >> 1;
+            } else {
+              diff = ~(diff >> 1);
+            }
+            ((unsigned int *)uncompressed_data)[beg + i] = diff + lastpix;
+            lastpix = ((unsigned int *)uncompressed_data)[beg + i];
+          }
+        }
+      }
+    }
+  }
+}
+
+void FitsReaderGPU::RunImpl(Workspace &ws) {
+  int num_outputs = ws.NumOutput();
+  int batch_size = GetCurrBatchSize();
+
+  for (int output_idx = 0; output_idx < num_outputs; output_idx++) {
+    auto &output = ws.Output<GPUBackend>(output_idx);
+    for (int sample_id = 0; sample_id < batch_size; ++sample_id) {
+      auto &sample = GetSample(sample_id);
+
+      if (sample.header[output_idx].compressed) {
+        void *decoded_data_cuda;
+        unsigned char *undecoded_data_cuda;
+        long *tile_offset_cuda;
+        long *tile_size_cuda;
+        int64_t tiles = sample.header[output_idx].tiles,
+                maxtilelen = sample.header[output_idx].maxtilelen,
+                zbitpix = sample.header[output_idx].zbitpix;
+
+        if (zbitpix == 8) {
+          cudaMalloc(&decoded_data_cuda, tiles * maxtilelen * sizeof(char));
+        } else if (zbitpix == 16) {
+          cudaMalloc(&decoded_data_cuda, tiles * maxtilelen * sizeof(short));
+        } else {
+          cudaMalloc(&decoded_data_cuda, tiles * maxtilelen * sizeof(int));
+        }
+
+        cudaMalloc(&tile_offset_cuda, (tiles + 1) * sizeof(long));
+        cudaMemcpy(tile_offset_cuda, sample.tile_offset[output_idx].data(),
+                   (tiles + 1) * sizeof(long), cudaMemcpyHostToDevice);
+
+        cudaMalloc(&tile_size_cuda, tiles * sizeof(long));
+        cudaMemcpy(tile_size_cuda, sample.tile_size[output_idx].data(), tiles * sizeof(long),
+                   cudaMemcpyHostToDevice);
+
+        cudaMalloc(&undecoded_data_cuda,
+                   sample.tile_offset[output_idx][tiles] * sizeof(unsigned char));
+        cudaMemcpy(undecoded_data_cuda, sample.data[output_idx].raw_data(),
+                   sample.tile_offset[output_idx][tiles] * sizeof(unsigned char),
+                   cudaMemcpyHostToDevice);
+
+        int blockSize = 256;
+        int numBlocks = (int)(tiles + blockSize - 1) / blockSize;
+
+        DALI_ENFORCE(false, "got here!");
+        rice_decompress<<<numBlocks, blockSize>>>(
+            undecoded_data_cuda, output.raw_mutable_tensor(sample_id), tile_offset_cuda,
+            tile_size_cuda, sample.header[output_idx].bytepix, sample.header[output_idx].blocksize,
+            tiles, maxtilelen);
+
+        cudaFree(undecoded_data_cuda);
```

Comment:
For tmp memory please use scratchpad - https://github.com/NVIDIA/DALI/blob/main/dali/operators/image/remap/remap.cu#L69 (declaration), https://github.com/NVIDIA/DALI/blob/main/dali/operators/image/remap/remap.cuh#L74 (allocates memory on the GPU  and copies data from the CPU to it using given stream).
Scratchpad should live until to the last call touching given memory. So:
```
dali::kernels::DynamicScratchpad ds;
auto stream = ws.stream();
auto tile_offset_cuda = std::get<0>(ds.ToContiguousGPU(stream, sample.tile_offset[output_idx]));
auto tile_offset_cuda = std::get<0>(ds.ToContiguousGPU(stream, sample.tile_size[output_idx]));
auto undecoded_data_cuda= std::get<0>(ds.ToContiguousGPU(stream, sample.data[output_idx]));
```